### PR TITLE
feat: support path in nGQL, optimized nGQL edge reader

### LINKED ISSUE #132: feat: ngql reader to support arbitrary query result parsing
Now only `MATCH ()-[e:xx]->() return e LIMIT 10000`is supported, ideally we could parse all edges of arbitrary results to enable complex `MATCH/FIND PATH/GET SUBGRAPH` to be leveraged, just add a whitelist of parsed edge schema.

### PR DESCRIPTION
**Pull request by wey-gu**
## What type of PR is this?
- [ ] bug
- [x] feature
- [ ] enhancement

#### Issue(s) number: 
close #132

#### Description:

We should not expect only edge or list of edges to be in resultSet for nGQL, this change fixes it.

## How do you solve it?

- Parse Relationship, too
- Fix spark3 missing OperateType
- Added examples and e2e tests for match/subgraph/find path with ngqlEdgeReader
- Added non-edge in List handling subroutine, i.e. subgraph return list of nodes could trigger exceptions that were unexpected before.

### CODE REVIEW

File: nebula-spark-connector/src/main/scala/com/vesoft/nebula/connector/reader/NebulaNgqlEdgePartitionReader.scala [hunk 1]
```diff
@@ -73,9 +73,15 @@ class NebulaNgqlEdgePartitionReader extends InputPartitionReader[InternalRow] {
...
+          val list: mutable.Buffer[ValueWrapper] = value.asPath().getRelationships()
+          edges.appendAll(
+            list.toStream
+              .filter(e => checkLabel(e.asRelationship()))
+              .map(e => convertToEdge(e.asRelationship(), properties))
+          )
         } else {
-          LOG.error(s"Exception convert edge type ${valueType} ")
-          throw new RuntimeException(" convert value type failed");
+          LOG.warn(s"Unexpected edge type ${valueType}. Skipping.")
```
Comment (Nicole00) [inline nebula-spark-connector/src/main/scala/com/vesoft/nebula/connector/reader/NebulaNgqlEdgePartitionReader.scala:84]: Connector is an API not a tool, why just skip but not throw exception?

Comment (wey-gu) [reply nebula-spark-connector/src/main/scala/com/vesoft/nebula/connector/reader/NebulaNgqlEdgePartitionReader.scala:84]: make sense, was doing so due to some of the users are not familiar, or experienced enough to inspect error logs properly.

Will bring back the exception.

File: nebula-spark-common/src/main/scala/com/vesoft/nebula/connector/NebulaConfig.scala [hunk 2]
```diff
@@ -83,6 +83,12 @@ object NebulaConnectionConfig {
     /**
       * set connectionRetry, connectionRetry is optional
       */
+    @deprecated("use withConnectionRetry instead", "3.7.0")
```
Comment (wey-gu) [inline nebula-spark-common/src/main/scala/com/vesoft/nebula/connector/NebulaConfig.scala:86]: @Nicole00 could we do so to correct the typo?

File: nebula-spark-connector/src/main/scala/com/vesoft/nebula/connector/reader/NebulaNgqlEdgePartitionReader.scala [hunk 3]
```diff
@@ -70,12 +70,19 @@ class NebulaNgqlEdgePartitionReader extends InputPartitionReader[InternalRow] {
           val list: mutable.Buffer[ValueWrapper] = value.asList()
           edges.appendAll(
             list.toStream
-              .filter(e => checkLabel(e.asRelationship()))
+              .filter(e => e.isEdge() && checkLabel(e.asRelationship()))
```
Comment (wey-gu) [inline nebula-spark-connector/src/main/scala/com/vesoft/nebula/connector/reader/NebulaNgqlEdgePartitionReader.scala:73]: This is helpful @Nicole00 to mitigate the case like NULL/EMPTY in List?

File: nebula-spark-common/src/main/scala/com/vesoft/nebula/connector/NebulaConfig.scala (hunk 2 above)
Comment (wey-gu) [reply nebula-spark-common/src/main/scala/com/vesoft/nebula/connector/NebulaConfig.scala:86]: maybe put 4.0.0/5.0.0 here instead?

Comment (wey-gu) [reply nebula-spark-common/src/main/scala/com/vesoft/nebula/connector/NebulaConfig.scala:86]: update, after the rebase I realized I was doing this on the wrong base, but we should maintain the capability of the interface for a while, so keep this old typo for a while. Or it'll break things

**Comment by wey-gu** [comment]
It's strange that after rebasing I kept got this error:

Screenshot 2023-11-22 at 17 10 56

```
- read edge from nGQL: MATCH ()-[e:friend]->() RETURN e LIMIT 1000 *** FAILED ***
  java.util.NoSuchElementException: key not found: operateType
  at scala.collection.MapLike$class.default(MapLike.scala:228)
  at org.apache.spark.sql.catalyst.util.CaseInsensitiveMap.default(CaseInsensitiveMap.scala:28)
  at scala.collection.MapLike$class.apply(MapLike.scala:141)
  at org.apache.spark.sql.catalyst.util.CaseInsensitiveMap.apply(CaseInsensitiveMap.scala:28)
  at com.vesoft.nebula.connector.NebulaOptions.<init>(NebulaOptions.scala:38)
  at com.vesoft.nebula.connector.NebulaDataSource.getNebulaOptions(NebulaDataSource.scala:147)
  at com.vesoft.nebula.connector.NebulaDataSource.createReader(NebulaDataSource.scala:40)
  at org.apache.spark.sql.execution.datasources.v2.DataSourceV2Relation$SourceHelpers.createReader(DataSourceV2Relation.scala:155)
  at org.apache.spark.sql.execution.datasources.v2.DataSourceV2Relation$.create(DataSourceV2Relation.scala:172)
  at org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:204)
  ...
```

File: nebula-spark-common/src/main/scala/com/vesoft/nebula/connector/NebulaConfig.scala (hunk 2 above)
Comment (Nicole00) [reply nebula-spark-common/src/main/scala/com/vesoft/nebula/connector/NebulaConfig.scala:86]: Agree! great idea.

**Review by Nicole00 — APPROVED** [review]
LGTM

**Pull request merged**